gobblin-data-management/src/main/java/org/apache/gobblin/data/management/source/LoopingDatasetFinderSource.java

```diff
@@ -85,53 +88,82 @@ public WorkUnitStream getWorkunitStream(SourceState state) {
       int maxWorkUnits = state.getPropAsInt(MAX_WORK_UNITS_PER_RUN_KEY, MAX_WORK_UNITS_PER_RUN);
```

Comment:
This method is getting a little hard to read. I suggest the following refactorings:

```
// (I suggest renaming maxRetryTimes to just maxRetries)
int maxRetries = getMaxRetries(state);

List<WorkUnit> failedPreviousWorkUnits = getFailedPreviousWorkUnits(previousWorkUnitStates);
WorkUnitState maxWorkUnitState = getMaxWorkUnitState(previousWorkUnitStates);

if (maxWorkUnitState != null && !maxWorkUnitState.getPropAsBoolean(END_OF_DATASETS_KEY, false)) {
  // set watermarks
}
```

I think it's worth iterating over previousWorkUnitStates multiple times for the sake of readability.